### PR TITLE
Added log-delivery-policy for s3-bucket component

### PR DESCRIPTION
**Pull request by zdmytriv**
## what
* Added log-delivery-policy for s3-bucket component

## why
* log-delivery-policy is common policy that is being added to S3 bucket. For example:
  * https://docs.aws.amazon.com/network-firewall/latest/developerguide/logging-s3.html
  * https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-query-logs-choosing-target-resource.html

### CODE REVIEW

**Review by Nuru — CHANGES_REQUESTED** [review]

File: modules/s3-bucket/main.tf [hunk 1]
```diff
@@ -10,7 +10,7 @@ locals {
 
   custom_policy       = var.custom_policy_enabled ? data.aws_iam_policy_document.custom_policy[0].json : null
   log_delivery_policy = var.log_delivery_policy_enabled ? data.aws_iam_policy_document.log_delivery_policy[0].json : null
-  bucket_policy       = var.custom_policy_enabled == var.log_delivery_policy_enabled ? data.template_file.bucket_policy.rendered : coalesce(local.log_delivery_policy, local.custom_policy)
+  bucket_policy       = coalesce(local.custom_policy, local.log_delivery_policy, data.template_file.bucket_policy.rendered)
 
   logging = var.logging != null ? {
     bucket_name = var.logging_bucket_name_rendering_enabled ? format(var.logging_bucket_name_rendering_template, var.namespace, var.tenant, var.environment, var.stage, var.logging.bucket_name) : var.logging.bucket_name
```
[the remark this answers is not in the document]
Comment (zdmytriv) [reply modules/s3-bucket/main.tf:16]: fixed

File: modules/s3-bucket/variables.tf [hunk 2]
```diff
@@ -379,14 +379,28 @@ variable "custom_policy_account_names" {
 }
 
 variable "custom_policy_enabled" {
-  description = "Whether to enable or disable the custom policy. If enabled, the default policy will be ignored"
+  description = <<-EOT
+  Whether to enable or disable the log delivery policy. Conflicts with `var.custom_policy_enabled` and `var.iam_policy_statements`".
```
[the remark this answers is not in the document]
Comment (zdmytriv) [reply modules/s3-bucket/variables.tf]: Oops you are right, fixed

File: modules/s3-bucket/variables.tf [hunk 3]
```diff
@@ -379,14 +379,28 @@ variable "custom_policy_account_names" {
...
   type        = bool
   default     = false
 }
 
 variable "iam_policy_statements" {
   type        = any
-  description = "Map of IAM policy statements to use in the bucket policy."
+  description = <<-EOT
+  Map of IAM policy statements to use in the bucket policy. Conflicts with `var.custom_policy_enabled` and `var.log_delivery_policy_enabled`.
+  It will be used if `var.custom_policy_enabled` and `var.log_delivery_policy_enabled` are set to `false`.
```
[the remark this answers is not in the document]
Comment (zdmytriv) [reply modules/s3-bucket/variables.tf:403]: nope.

When `custom_policy_enabled=true` policy from `iam.tf` [this](https://github.com/cloudposse/terraform-aws-components/pull/698/files#diff-37adfa2463603828781680ac4b93d07be0be369e8aed2a393f56784733b3cd08R1-R18) will be used.

When `iam_policy_statements=true` (and other 2 are `false`) policy from `iam-policy` module will be used [this](https://github.com/cloudposse/terraform-aws-components/pull/698/files#diff-0f985f94d08f26bd28e34176ac4fd45f37a344c744703939cc33217a6a843a96R25-R42)

File: modules/s3-bucket/iam.tf [hunk 4]
```diff
@@ -0,0 +1,52 @@
+data "aws_iam_policy_document" "custom_policy" {
+  count = local.enabled && var.custom_policy_enabled ? 1 : 0
+
+  statement {
+    actions = var.custom_policy_actions
+
+    resources = [
+      format("arn:%s:s3:::%s", local.aws_partition, module.this.id),
```
[the remark this answers is not in the document]
Comment (zdmytriv) [reply modules/s3-bucket/iam.tf]: I was also wondering why it is done this way but didn't want to change extra not related to this PR stuff.

Fixed

**Comment by zdmytriv** [comment]
@Nuru please re-review

**Review by Nuru — CHANGES_REQUESTED** [review]
@zdmytriv I appreciate the motivation behind this PR and the effort you put into it. I am, however, not inclined to accept this PR for a few reasons. (Although I'm using the "Request changes" feature, I'm not asking you to make changes to make the PR acceptable, I'm asking you to close the PR without merging.)

To begin with, the log delivery policy you have added is not the [current "best practice" policy](https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-s3-permissions.html). The current best practice is to include conditionals on the source account and ARN to guard against the confused deputies problem. (See the link for more details.)

Next, the name "custom policy" of the input to the current component is a bit misleading. Both the custom policy and the default policy are user supplied policies, just with different formats. This component does not provide any pre-defined policies. Adding a pre-defined policy makes this component even more complicated, and invites future users to want to be able to combine custom and pre-defined policies in a way that will cause further confusion and headache.

Cloud Posse already provides other components and modules for log delivery buckets, such as `vpc-flow-logs-bucket` and `cloudtrail-bucket` to handle the cases where the S3 bucket is for log delivery. These are much easier to use because they are pre-tuned for specific purposes. This component (s3-bucket) is meant for maximum flexibility, and users can and should create their own policies for it.

So I suggest you use one of the other components that meets your needs and please close this PR.

**Pull request closed without merging**